File: myLeetCode/Hash Table/1002. Find Common Characters.cpp
// 1002. Find Common Characters
// https://leetcode.com/problems/find-common-characters/

#include <iostream>
#include <vector>
#include <climits>

class Solution {
public:
    std::vector<std::string> commonChars(std::vector<std::string>& words) {
        std::vector<int> minfreq(26, INT_MAX);
        std::vector<int> freq(26);
        for (const std::string& word: words) {
            std::fill(freq.begin(), freq.end(), 0);
            for (char ch: word) {
                ++freq[ch - 'a'];
            }
            for (int i = 0; i < 26; ++i) {
                minfreq[i] = std::min(minfreq[i], freq[i]);
            }
        }

        std::vector<std::string> ans;
        for (int i = 0; i < 26; ++i) {
            for (int j = 0; j < minfreq[i]; ++j) {
                ans.emplace_back(1, i + 'a');// emplace_back(n,m)尾端插入n个m
            }
        }
        return ans;
    }
};

int main() {
    Solution sol;
    std::vector<std::string> words = {"bella","label","roller"};
    std::vector<std::string> ans = sol.commonChars(words);
    for (auto& str: ans) {
        std::cout << str << " ";
    }
    std::cout << std::endl;
    return 0;
}

File: myLeetCode/Hash Table/1002. Find Common Characters.py
# 1002. Find Common Characters
# https://leetcode.com/problems/find-common-characters/

class Solution:
    def commonChars(self, words: list[str]) -> list[str]:
        minfreq = [float('inf')] * 26
        for word in words:
            freq = [0] * 26
            for ch in word:
                freq[ord(ch) - ord('a')] += 1
            for i in range(26):
                minfreq[i] = min(minfreq[i], freq[i])
        ans = []
        for i in range(26):
            ans.extend([chr(i + ord('a'))] * minfreq[i])
        return ans
    
def main():
    sol = Solution()
    words = ["bella","label","roller"]
    ans = sol.commonChars(words)
    print(ans)

if __name__ == "__main__":
    main()